usr/sbin/pkcsicsf/pkcsicsf.c

```diff
@@ -136,46 +138,27 @@ int get_pin(char **pin, size_t *pinlen)
     return rc;
 }
 
-int get_start_slot(void)
+static int get_start_slot(struct ConfigBaseNode *config)
 {
-
-    FILE *fp;
-    char temp[LINESIZ];
-    int num;
-    int found = -1;
-    struct stat statbuf;
-
-    /* if file doesn't exist, then first slot will be 0. */
-    if ((stat(OCK_CONFIG, &statbuf) < 0) && (errno == ENOENT))
-        return 0;
-
-    fp = fopen(OCK_CONFIG, "r");
-    if (fp == NULL) {
-        fprintf(stderr, "open failed, line %d: %s\n",
-                __LINE__, strerror(errno));
-        return -1;
-    }
-
-    /* step thru config file and get biggest slot number used */
-    while (fgets(temp, LINESIZ, fp) != NULL) {
-        if (strstr(temp, SLOT) != NULL) {
-            if (sscanf(temp, "%*s %d", &num) == 1)
-                if (num > found)
-                    found = num;
+    struct ConfigBaseNode *c;
+    struct ConfigIdxStructNode *slot;
+    int i, slot_id = -1;
+
+    confignode_foreach(c, config, i) {
+        if (confignode_hastype(c, CT_IDX_STRUCT)) {
+            slot = confignode_to_idxstruct(c);
+            if (strcmp(slot->base.key, "slot") == 0) {
+                if ((int)slot->idx > slot_id)
+                    slot_id = slot->idx;
+                continue;
+            }
         }
     }
 
-    /* bump up the slot number, since this will be next new slot entry.
-     * if it was an empty file or a file with no slot entries,
-     * then next new slot entry will be 0.
-     */
-
-    fclose(fp);
-
-    return ++found;
+    return ++slot_id;
```

Comment:
I still think this is dangerous.  Even if I only have one slot, but this slot has the maximum number, then this tool will produce an invalid configuration.  Personally, I think we should go for a two-pass here: first pass mark in a 1024-byte array all slots that are actually used; second pass search in the array for an unmarked slot.

Reply:
Well, the problem is that the calling routine config_add_slotinfo() uses the value returned by get_start_slot() as a start value, and may increase it for further slots it adds. So even if we find a free slot id, we can't guarantee that subsequent slots are unused. So I guess the logic must be changed to let config_add_slotinfo() call get_start_slot() every time to get the next free slot-id (instead of simply incrementing it). I will then rename get_start_slot() to get_free_slot()....

Reply:
In that case I would suggest to compute the bitmap in this function and write a function to search the first unset bit in the map for every slot that gets added.  Yes, quite a change to the old version, but it is the closest thing to correct I can think of.